Feat command mode

NickHackman opened the pull request:
### Features
 * infobar displays command mode input, and errors
#### Command Mode Features
     * h/help
     * q/quit
     * sort
     * ng/newglob, handles `~` and `$HOME` properly and directories are assumed to be `/dir/*`
     * df/destfolder
     * r/reverse
     * m/max

### Known bugs: 

* In command mode: cannot type `:` due to the event buffer not being cleared this character is ignored, because it will commonly be added on first loop of `get_command` in `/src/program/command_mode.rs`

* When program starts in a directory that has no images, Command mode isn't properly displayed on the infobar

### Future Fixes
 * Decouple the rendering so that I and @gurgalex can update the infobar without having to pass the info to display to `render_screen`, allow updating of smaller components like the infobar before calling render or some other solution.

* Force rendering, workaround solution being used for cases like `newglob` which should re-render when the current viewed image isn't in the new glob.

@gurgalex Here's how I'm handling it, few bugs I wanted to iron out before setting up the pull request

NickHackman committed:
Add vim like command mode

* Colon now enters command mode
* Added a modal approach kept track of in Program struct
    Current modes: normal, command, exit
* Changed previous fn run -> fn run_normal_mode
* Add basis for future command mode commands

NickHackman committed:
Add cmd mode for infobar, cmd mode help and quit

* Modified infobar so that it takes a &str called msg in the case of
    command mode displays what the user is currently typing
* Add :help to print the help message and return to normal mode
* Add :quit to terminate the program from command mode

NickHackman committed:
Add Command mode

Command mode is any mode that requires user input

 * Moved command_mode related functions into its own file
 * Command Mode Features:
    * newglob (takes only one paramter). Properly handles ~/$HOME when
	provided by user. When the current path exists in the current vec of images
	and in the future one, remain at its index

    * sort (optionally one paramter). if called alone just performs the
	selected sort on the images, if provided with a second argument (a
	SortOrder) switch to that then sort. User is moved to new index
	of current image prior to sort

    * reverse, reverses the current set of images, keeping the index of
	the current image

    * destfolder, sets the destination folder

    * help, exits command mode and displays help info in normal mode

    * quit, terminates program

 * info bar displays user input, and that the user is in command mode
 * window event handling while in command mode

NickHackman committed:
Add command mode m/max, takes new max image length

NickHackman committed:
Refractored code, fixed bug in :max

 * Moved the event match and Mode to ui.rs
 * Fixed bug, where input for :max could be 0 and it would be
    interpreted as 0 instead of infinitely many

NickHackman committed:
Add force render, added docs to command_mode_run

NickHackman committed:
Refractor name change in ui.rs

NickHackman committed:
Add newglob Err msg if file doesn't exist, bug fix

Increased performance, inside command_mode function get_command by not
rendering user input ever event, but only when it changes

gurgalex commented on src/program/command_mode.rs:
```diff
@@ -0,0 +1,263 @@
+                    return Err(err_msg);
+                }
+                let mut new_images: Vec<PathBuf>;
+                new_images = glob_path(input_vec[1])?;
```
This would fail if the glob had a file or directory with a ` ` in it or escaped space even `\ `.

Example
`newglob /home/alex/My\ Pictures/Tour\ S*.png`

NickHackman replied:
Good catch! I completely forgot about that edge case, thanks! :+1:

What do you think about the implementation of command mode as a whole? Is there a better approach that I'm missing?

gurgalex replied:
I like this implementation better. (I mostly just want something that seems to work).

I do like that text entry doesn't depend on keycodes. (Hello, international symbols 「東方」).

Does this string expand out `ng ~/*.png`?

Shellexpand may reduce some of the hand-written code for ~ expansion.
http://docs.rs/shellexpand/1.0.0/shellexpand/

gurgalex replied:
I really was thinking we'd need to do something for custom commands using

https://pest.rs/

or nom
https://github.com/Geal/nom/


Maybe eventually, though it'd be great? if we didn't have to.

NickHackman replied:
> Does this string expand out `ng ~/*.png`?
> 
> Shellexpand may reduce some of the hand-written code for ~ expansion.
> http://docs.rs/shellexpand/1.0.0/shellexpand/

As of right now that doesn't work, but I think that's more of an issue with `glob` than us. I'll look into `shellexpand` thanks :smiley:

gurgalex replied:
nice live grammar evaluator they have
Nice tools and documentation if needed though.

https://pest.rs/#editor

And a book
https://pest.rs/book/

nom still an option
https://crates.io/crates/nom

NickHackman replied:
@Davejkane, @gurgalex potential futures issues based off of work done in command mode and discussed in #57

### Future Feature Issues
 * Infobar is easier to update/decoupled from the rest of rendering or refractored in a way where smaller components of rendering can be updated or rendered without rerendering everything
 * **Good First Issue** Info bar accepts the input of error messages, changes the current image index field to a red color and displays "Error" and then the error message in the current image path field. Update errors to now, if possible, be sent to the infobar removing "Error:" when necessary from the message
* **Good First Issue** Add the `add` command to command mode, which should follow the format `:add glob file directory` or `:a glob file directory` and append images to `self.paths.images`
* **Good First Issue** Add the `delete` command to command mode, which should follow the format `:delete /path/to/image regex` or `:d /path/to/image` otherwise delete current image from `self.paths.images`. *Don't actually delete the file*

Any other ideas? I feel like these are good first issues that require some work, but not like a crazy amount where they need to understand how the app functions as a whole.

NickHackman replied:
So I've looked into http://docs.rs/shellexpand/1.0.0/shellexpand/ and it uses [std::env::home_dir](https://doc.rust-lang.org/std/env/fn.home_dir.html) which is deprecated, Opened an [issue](https://github.com/netvl/shellexpand/issues/3). Afterwards it may be useful though!

NickHackman commented:
@Davejkane, @gurgalex potential future issues based off of work done in command mode and discussed in #57

### Future Feature Issues
 * Infobar is easier to update/decoupled from the rest of rendering or refractored in a way where smaller components of rendering can be updated or rendered without rerendering everything
 * **Good First Issue** Info bar accepts the input of error messages, changes the current image index field to a red color and displays "Error" and then the error message in the current image path field. Update errors to now, if possible, be sent to the infobar removing "Error:" when necessary from the message
* **Good First Issue** Add the `add` command to command mode, which should follow the format `:add glob file directory` or `:a glob file directory` and append images to `self.paths.images`
* Add autocomplete for command mode for files, commands, etc
* Force rendering of application as mentioned above
* **Good First Issue** Keep a history of command mode commands entered by user

Any other ideas? I feel like these are good first issues that require some work, but not like a crazy amount where they need to understand how the app functions as a whole.

edit: moved from a sub comment into main discussion.

NickHackman committed:
Add Commands enum for command mode

* Enum that lists out the possible commands and implements FromStr for
each

gurgalex commented:
> Good First Issue Add the delete command to command mode, which should follow the format :delete /path/to/image regex or :d /path/to/image otherwise delete current image from self.paths.images. Don't actually delete the file

Maybe `untrack`?. I'd like to reserve the word `delete` or `remove` for actually deleting an mage from storage. 
Yes, the fn name currently is called `remove_image`, so possibly need to rename that sometime.

NickHackman committed:
Fix bug in newglob, add dep shellexpand and regex

* Newglob wouldn't properly accept files like /home/etc/test\ file\ name
* Add dependency regex to be able to fix Newglob bug
* Add dependency shellexpand to properly handle ~/$HOME/env vars
* Minor refractors

NickHackman committed:
Fixed Bugs, empty images didn't render bar command

* Render self.render_blank now passes command mode input to it to
properly render, so the user can populate the images
* Fixed bug, in newglob where it would crash on empty self.paths.images

NickHackman commented:
Update on progress:
* Added dependency [shellexpand](https://github.com/netvl/shellexpand) to better expand `~` and `$HOME` and other env variables. Note `shellexpand` uses deprecated `std::env::home_dir` pull request sent in fixing that.
* Fixed bug, where command bar wouldn't be rendered on empty images
* Fixed bug, mentioned by @gurgalex where newglob wouldn't accept paths like `/home/etc/path\ to\ dir/*` by adding dependency [regex](https://github.com/rust-lang/regex) (will be used later on for search as well)
* Overall, restructure of code base in `src/program/command_mode.rs` to be more intuitive

Opening code for review

NickHackman committed:
Merge branch 'development' into feat-redefine-glob

NickHackman committed:
Refractored code in response to clippy suggestions

NickHackman committed:
Merge remote-tracking branch 'refs/remotes/origin/feat-redefine-glob' into feat-redefine-glob

Davejkane commented on src/program/command_mode.rs:
```diff
@@ -0,0 +1,300 @@
+/// TODO: have regex capture the ending space in case of escaped files
+fn parse_user_input(input: &str) -> Vec<String> {
+    let mut input_vec: Vec<String> = Vec::new();
+    let re: Regex = Regex::new(r"[\x21-\x7E]+(\\)?").unwrap();
```
Please add a comment to the line. I think it's fair to say it's not self-explanatory.

NickHackman replied:
:+1:

Davejkane commented on src/program/command_mode.rs:
```diff
@@ -0,0 +1,300 @@
+        }
+    }
+    input_vec
+}
```
I'm not entirely clear on what the above function does. Is it for parsing user input, specifically for the path + glob, or something else? Why is it limited to ascii? You should probably mention that in the function name, that it is specific to paths. Also, guess this won't work on windows?

NickHackman replied:
Added better documentation, but you're completely right my code base isn't cross platform.

This functionality will come in the next push or so :heart:

NickHackman replied:
Fixed in latest push, completely OS agnostic

marking as resolved

Davejkane commented on src/infobar.rs:
```diff
@@ -6,29 +6,44 @@ use crate::paths::Paths;
+            Some(msg) => {
+                // user is in command mode
+                information = msg.to_string();
+                mode = String::from("Command");
```
Am I right in thinking that if I send Some() string to this function, then I'll end up in Command mode? Will this effect using the infobar for error messages?

NickHackman replied:
Yeah, this was the main issue addressed in "Rewrite of Modal system".  New changes
```rust
pub enum Mode {
    /// Default mode, allows the removal, traversal, move, and copy of images
    Normal,
    /// Mode that is built off of user input, allows switching the current glob
    /// string is the input to display on the infobar
    Command(String),
    /// Mode that is meant to display errors to the user through the infobar
    /// string is the input to display on the infobar
    Error(String),
    /// Terminate condition, if this mode is set the program will stop execution
    Exit,
}
```
The `render_infobar` funciton now passes in the current mode, which contains the message if applicable and This required me to change a lot of the way I handled error handling, so now Error was added as an easy solution.

Davejkane commented on src/program/command_mode.rs:
```diff
@@ -0,0 +1,300 @@
+impl<'a> Program<'a> {
+    /// User input is taken in and displayed on infobar, cmd is either '/' or ':'
+    /// Returning empty string signifies switching modes back to normal mode
+    // TODO: autocomplete use fn
```
I'm not a big fan of TODOs left in the code. Can we please deal with the things that can be dealt with and make issues for the things that need to be follow-ups.

NickHackman replied:
Fixed in "Rewrite of Modal system"

Davejkane commented on src/ui.rs:
```diff
@@ -105,13 +126,20 @@ pub fn event_action(state: &mut State, event: &Event) -> Action {
-            Home => state.process_action(Action::First),
-            End => state.process_action(Action::Last),
+            Semicolon => {
+                if state.left_shift || state.right_shift {
```
Really not sure this holds true on non-american keyboards. Shift semicolon does not always equal colon. Please see if there's another way to do this. Maybe a virtual keycode if you can get them working.

Davejkane commented on src/ui.rs:
```diff
@@ -105,13 +126,20 @@ pub fn event_action(state: &mut State, event: &Event) -> Action {
+            W | PageUp => Action::SkipForward,
+            B | PageDown => Action::SkipBack,
+            Z => Action::ToggleFit,
+            Home => Action::First,
```
Why are we no longer calling `process_action` on these Actions?

NickHackman replied:
Fixed in "Rewrite of Modal system, add Error"

Thanks for catching that, that was completely unintended. This was a mistake in my merging, my bad.

Davejkane reviewed (changes requested):
I've made a couple of comments, nothing too major. In general I'm impressed with the code and certainly with the enthusiasm to just jump in and make such a huge feature. The project is going to be all the better for it.

NickHackman committed:
Rewrite of Modal system, add Error

This solution path was taken in response to the sketchy passing command
to render_screen. This is a more proactive approach that is easier to adapt in the future.

* Add Mode, Error, to display error messages
* ui.rs Mode variants now take Strings if they need to display on infobar
* Fixed documentation, removed TODOs
* Infobar::update now matches on Modes

gurgalex commented on src/program/command_mode.rs:
```diff
@@ -98,26 +97,32 @@ fn glob_path(path: &str) -> Result<Vec<PathBuf>, String> {
+    // Regex pattern matches all unicode characters there must be at least 1
+    // and the '\' character may be present at the end of the string
+    // This is used to rather than split on whitespace, only get the actual arguments
+    let regex = match Regex::new(r"[\x21-\x7E]+(\\)?") {
```
I believe this regex should be compiled only once, and not on every function call.

https://docs.rs/regex/1.1.6/regex/#example-avoid-compiling-the-same-regex-in-a-loop
Plus we could eventually have a descriptive name for every regex made.

NickHackman replied:
Yep, with the cross platform approach I'm thinking about this would be a global constant in `command_mode.rs`. 

Coming up with really good names is always hard...

gurgalex replied:
So, how would I type this as newglob?
`/home/alex/test dir/**/*.png`

would it be
`ng /home/alex/test\ dir/**/*.png`?

NickHackman replied:
Yeah, because otherwise it's impossible to know the end of a file name without making syscalls for every string provided and their combinations.

Funny enough, this is a bug in Neovim, and I'm not making an issue because it's impossible to fix...

make a file `~/home/$USER/test\\`
open neovim
:e test\\\ Documents/*
current buffer: `test\\ Documents/*`

This results in whenever you try to save the current buffer it fails the write operation because that's never allow to be a file which is just awesome. 

In our case we don't have that bug because of cli.rs/push_image_path just gives back nothing...

gurgalex replied:
![gotten_glob](https://user-images.githubusercontent.com/28968644/58051234-41ce1580-7b17-11e9-87dd-62b75215deeb.png)

gurgalex replied:
Does `ng` with a directory with spaces work for you?

NickHackman replied:
It does not, actually need to remove the `\`s in order for it to be recognized as a dir. Thanks for the catch, I completely neglected that.

Davejkane replied:
path_regex would be a fine name, and of course, it should only be compiled once, even though the overhead is presumably tiny.

NickHackman replied:
Fixed, via a large amount of rewriting 

A regex was an over complicated approach that was naive and shortsighted, for instance our input to command mode may not always a be a file. Each command should handle and parse its own input since it can be interpreted differently in this context. 

Marking as resolved.

gurgalex reviewed (changes requested):
Avoid recompiling the regex every time it is called. See https://docs.rs/regex/1.1.6/regex/#example-avoid-compiling-the-same-regex-in-a-loop

NickHackman commented:
@Davejkane thanks for the advice/review! in the future I'm going to have a more cross platform mindset compared to currently as that seems like a big issue. In addition to that a more proactive mindset as well some of the code I wrote in this was quick and dirty/sketchy. I'm going to iron out these issues ASAP :smiley: 

Yeah... this seemed like it wouldn't be too bad going in, but quickly became this behemoth.

I think that we should remove all the normal mode features from `src/program/mod.rs` -> `src/program/normal_mode.rs` or something along those lines to better layout the project, just a thought though

NickHackman committed:
Refractored code in command_mode.rs

Davejkane commented on src/ui.rs:
```diff
@@ -126,17 +130,16 @@ pub fn process_normal_mode<'a>(state: &mut State<'a>, event: &Event) -> Action<'
+            W | PageUp => state.process_action(Action::SkipForward),
+            B | PageDown => state.process_action(Action::SkipBack),
+            Z => state.process_action(Action::ToggleFit),
             Home => Action::First,
```
Please process the action for all the actions. Otherwise if some first deletes an image, then goes to the first image, then uses the dot operator, they will delete another image, and they're probably not expecting that.

Davejkane reviewed (changes requested):
Still a few little things. I would like us to try to support windows if at all possible. At some point it would be nice to get a windows user on board. I have no idea if this even works on windows.

gurgalex committed:
Add lazy_static!. Make path_regex compile once

gurgalex committed:
Remove unneeded lifetimes due to cloning

NickHackman committed:
Refractored command_mode.rs

* Removed regex dependency
* changed how user input is parsed

NickHackman committed:
Merge remote-tracking branch 'refs/remotes/origin/feat-redefine-glob' into feat-redefine-glob

* Removed dependency regex and lazy_static

NickHackman committed:
Fixed issue with merge

NickHackman committed:
Fixed bug in sort on empty, and when merging

* Sorting on empty would panic as a result of trying to find the current
image, prior to sort in images
* Bug that was introduced during merging

Davejkane commented:
As far as I'm concerned, there's just a couple of things left. And that's the "colon is not necessarily shift+semi-colon" issue. And then there were some actions that appeared to be unprocessed and so I'm not sure if they would apply to the dot operator, but it could be that I've misunderstood a change there.

gurgalex commented:
@Davejkane  @NickHackman 
Layout independent `:` could be resolved with a mixture of keycodes and scancodes.

Use keycodes for detecting when the OS generates a `LSHIFTMOD` or `RSHIFTMOD`. Match the scancode for a `;` on a QWERTY keyboard. 

[KeyDown event struct](https://rust-sdl2.github.io/rust-sdl2/sdl2/event/enum.Event.html#variant.KeyDown)

From [sdl2 docs](http://wiki.libsdl.org/CategoryKeyboard)
>    SDL_Scancode values are used to represent the physical location of a
> keyboard key on the keyboard.
>
>    SDL_Keycode values are mapped to the current layout of the keyboard and
> correlate to an SDL_Scancode.

>  Which one to use is left to the application: scancodes are suited in
> situations where controls are layout-dependent (eg. the "WASD" keys as
> left-handed arrow keys), whereas keycodes are better suited to situations where
> controls are character-dependent (eg. the "I" key for Inventory).

NickHackman commented:
@Davejkane Little confused what you're referring to, what operations were being unprocessed?

@gurgalex Thanks

Thanks so much for the input it has helped monumentally, and sorry that this has taken so long :heart:

Davejkane commented:
I'm not sure I'm with you there @gurgalex. The problem isn't that semi-colon is sometimes in a different place, the problem is that colon is not always shift+semi-colon. For example, on a French keyboard (AZERTY) it is an unshifted key, whereas on a German keyboard (QWERTZ), it is shift+comma. So I'm not sure what you are suggesting would work.

NickHackman commented:
Yeah, going to use [scancode](https://docs.rs/sdl2/0.32.2/sdl2/keyboard/enum.Keycode.html#method.from_scancode) which uses the virtual mapping, going to convert that then check to see if it matches afterwards. I think that's keyboard independent.

Davejkane commented:
@NickHackman I see that I've misunderstood something there with the unprocessed actions. So just ignore that comment. Just the colon key to handle and then I'm happy.

gurgalex commented:
@Davejkane 
>I'm not sure I'm with you there @gurgalex. The problem isn't that semi-colon is sometimes in a different place, the problem is that colon is not always shift+semi-colon. For example, on a French keyboard (AZERTY) it is an unshifted key, whereas on a German keyboard (QWERTZ), it is shift+comma. So I'm not sure what you are suggesting would work.

Right, so scancodes return the QWERTY key which corresponds to the physical location of the key pressed regardless of OS keyboard mapping. So, a keypress on an AZERTY or QWERTZ in the same physical location will return the same scancode. (Assuming the hardware mapping isn't different)

Davejkane commented:
Actually, one last thing! If the bar is hidden because it is toggled off with the `t` key, then entering command mode should force it to be rendered.

Davejkane commented:
Also, something has been broken with move and delete image. The length of the images vector isn't updated, so the program crashes when moving to the last image. I'm going to have to do a bit more QA on this to see if anything else is broken.

gurgalex commented:
@NickHackman 

the following results in a panic.

Enter command mode (:)

type `ng` (no space)

Press Enter

```
thread 'main' panicked at 'byte index 3 is out of bounds of `ng`', src/libcore/str/mod.rs:2010:9
```

Davejkane commented:
@gurgalex, I'm still not sure I understand what you mean. Are you suggesting that a German user types in `Ö` to enter command mode? Because that would be what they would be doing to use the key that is physically in place where an American keyboard has semi-colon.

gurgalex commented:
>  Are you suggesting that a German user types in Ö to enter command mode? Because that would be what they would be doing to use the key that is physically in place where an American keyboard has semi-colon.

@Davejkane 
Yes.

Davejkane commented:
Also, @NickHackman no stress and no rush at all. It's totally fine and normal that things break on a big change like this.

Davejkane commented:
@gurgalex I think a German should type colon like everyone else to get into command mode, otherwise we'll be making lots of user manuals.

gurgalex reviewed (changes requested):
Fix crash outlined in https://github.com/Davejkane/riv/pull/58#issuecomment-494526131

NickHackman commented:
@gurgalex That's not reproducible on my end. Maybe pull again, that was an issue created by me messing up the merging, I fixed that in the latest push.

gurgalex commented:
> @gurgalex I think a German should type colon like everyone else to get into command mode, otherwise we'll be making lots of user manuals.

@NickHackman 
Okay, in that case. Matching on `TextInput` {text: ":", ..} should work instead.

https://docs.rs/sdl2/0.32.2/sdl2/event/enum.Event.html#variant.TextInput

gurgalex commented:
@NickHackman
You're right. I wasn't using the latest changes.

> @gurgalex That's not reproducible on my end. Maybe pull again, that was an issue created by me messing up the merging, I fixed that in the latest push.

NickHackman committed:
Added virutal mapping for ':'

* Support for shift semicolon as colon
* Support for colon on its own
* Support for virtually mapped keys to colon

NickHackman commented on src/ui.rs:
```diff
@@ -120,7 +154,17 @@ pub fn event_action(state: &mut State, event: &Event) -> Action {
-            _ => Action::Noop,
+            _ => {
+                // convert scancode to keycode via virtual mapping
+                let keycode = match sdl2::keyboard::Keycode::from_scancode(*scan) {
```
Added virtual mapping handled here, but this doesn't actually work on my end, the keycode produced is "semicolon"

So now there is handling for semicolon, colon, and virtual mapping
as mentioned [here](https://wiki.libsdl.org/SDL_Keycode) this just says that the colon isn't physically mapped

Davejkane replied:
What about plain old colon as keycode https://rust-sdl2.github.io/rust-sdl2/sdl2/keyboard/enum.Keycode.html#variant.Colon . Def doesn't work?

NickHackman replied:
Yeah, that doesn't work on my end nor does `KpColon` wish Colon worked though

Davejkane replied:
text input does seem like an ugly hack, but it might be the best bet. FYI, I was too lazy to handle this earlier, which is why I went with H for help instead of question mark like I originally planned. So thanks for taking this on.

NickHackman replied:
Okay, I'm on it

Davejkane replied:
Hey, and seeing as you have an open pull request, can you please put the stuff about sorting below the stuff about basic controls in the README :)

NickHackman reviewed (commented):
This feels like a bad approach, but it covers all the possible bases. It would have weird side effects, as mentioned above by @Davejkane `Ö` would also enter you into command mode, which is undesirable... 

Is the only approach that works is changing everything to be interpreted as `text_input`?

Davejkane commented:
So sorry @NickHackman, but the command for changing the destination folder isn't handling tilde properly. It creates a folder in the current directory called `~` which is difficult to `cd` into 😂

NickHackman committed:
Changed Event Handling for Normal mode

* sdl2 text_input is used to handle keys like 'j','k', 'q', etc
* changed help to '?' instead of 'h'

NickHackman committed:
Fixed bug max_viewable, would be overwritten

Max_viewable wouldn't be kept track of if in a glob that had a smaller
length when switching to one that would've exceeded the max_viewable set

NickHackman committed:
Updated README with Command mode options

NickHackman commented:
@Davejkane oh dang, my bad, fixing that asap

NickHackman committed:
Fixed bug in command mode destfolder

* Now properly handles env vars and ~

gurgalex commented on src/ui.rs:
```diff
@@ -40,8 +48,23 @@ pub enum Action {
 /// State tracks events that will change the behaviour of future events. Such as key modifiers.
-pub struct State {
+pub struct State<'a> {
     /// left_shift tracks whether or not the left shift key is pressed.
```
We can probably get rid of tracking `left_shift` and `right_shift` now that textinput will give the character ex. "G"

gurgalex commented on src/ui.rs:
```diff
@@ -72,46 +97,52 @@ impl State {
             Period => state.last_action.clone(),
-            Home => state.process_action(Action::First),
-            End => state.process_action(Action::Last),
             LShift => {
```
Removing `LShift` and `RShift` here as well.

NickHackman committed:
Removed unecessary L/R shift in ui.rs

gurgalex commented:
@NickHackman 

Command mode does not trigger if the infobar is set to hidden.

We may need to have two bars (an infobar and a command bar).

The command bar renders over the infobar (or just the image if hidden).

gurgalex commented:
Nevermind, it does trigger when hidden. Just was confused about what I had inputted (errors were not shown in input due to hidden bar)

Suggestion about a command bar still may be useful for showing input and errors.

NickHackman committed:
Entering command mode toggles bar on by default

NickHackman commented:
@gurgalex now it for sure is visible, just in case.

gurgalex commented:
@Davejkane @NickHackman 
Do you think equivalent command mode operations should be repeatable with `.` (such as `:help`) -> Action::Help

gurgalex commented:
Vim seems to use `@:` for repeating the last operation in command mode.
https://vim.fandom.com/wiki/Repeat_last_change

gurgalex commented:
We could address this later though.

NickHackman commented:
@gurgalex that sounds like maybe a different PR, but I like the idea

Davejkane commented:
@NickHackman wow, I'm impressed you got through so many tweaks. One of the big ones left is still that when I move or delete an image, the len of the images vector isn't being updated, meaning that then visiting the last file causes an out of bounds panic.

NickHackman committed:
Fixed bugs delete, move, and command mode df

* delete and move would panic when deleting/moving from the end or on
empty
* command mode destfolder/df wouldn't remove escaped spaces

NickHackman commented:
@Davejkane I don't think I wrote that bug, but fixed :+1:

edit: Yeah, sorry that is certainly my fault

Davejkane commented:
It was working fine before the new `max_viewable` logic. And as far as I can tell, it isn't fixed. I just tried to delete the last image when I had 4 images, and it crashed the program.

NickHackman commented:
@Davejkane I'm confused, that's not reproducible on my end. 
You updated your local repo after the push right?

Davejkane commented:
OK, just tried again and that seems to be fine, think pulling the branch didn't work. The program now seems to be exiting with this error message now `Error: "SDL2 error: Text has zero width"` when I hit `q` or `esc`. I promise that's my last gripe.

gurgalex commented on src/program/render.rs:
```diff
@@ -104,12 +105,12 @@ impl<'a> Program<'a> {
 
     fn render_infobar(&mut self) -> Result<(), String> {
-        let text = infobar::Text::from(&self.paths);
+        let text = infobar::Text::update(&self.ui_state.mode, &self.paths);
```
@NickHackman 

Here is where the code errors out.

This debug code shows that the `information` and `mode` are empty strings when the program starts to exit.

```rust
    fn render_infobar(&mut self) -> Result<(), String> {
        let text = infobar::Text::update(&self.ui_state.mode, &self.paths);
        dbg!((&text.mode, &text.information));
        // Load the filename texture
        let filename_surface = self
            .screen
            .font
            .render(&text.information)
            .blended(text_color())
            .map_err(|e| e.to_string()).expect("80: error");

```

gurgalex commented on src/infobar.rs:
```diff
@@ -3,32 +3,52 @@
+                (mode, information)
+            }
+            Mode::Error(msg) => ("Error".to_string(), msg.to_string()),
+            _ => (String::new(), String::new()),
```
@NickHackman 
(50) This is the line.

The program renders the infobar for the `Exit` mode which returns an empty string for both `mode` and `information`

NickHackman committed:
Fixed bug, program returns an error

* Infobar was rendering empty Strings for the Exit mode

NickHackman committed:
Command mode newglob, reset keep folder

* If destfolder is still 'keep' change it whenever newglob changes
directories

NickHackman commented:
Honestly guys thank you so much for all the help, I didn't handle so many edge cases that would've come back to haunt me in the near future. I think we ironed at least a large majority of them, can never say they're all gone, but all the ones that I've tested for are gone. I'm honestly really happy with the result and the level of code quality :smile: 

It was a joy working with you both on this, and I hope to contribute to the project more in the future :heart:

Probably going to do #53 after this

gurgalex commented on src/program/command_mode.rs:
```diff
@@ -310,6 +335,9 @@ impl<'a> Program<'a> {
                 }
                 self.newglob(&arguments);
+                if self.paths.dest_folder.ends_with("keep") {
+                    self.paths.dest_folder = self.paths.current_dir.join("keep");
```
@NickHackman 

Sorry to request more changes.

The destination folder should be kept the same unless the user changes it with `df`

I believe the field `current_dir` should be called `base_dir` to say that this is the base directory from which the glob finds images.

NickHackman replied:
In your above example that's how it should function currently. 

The only instance it would change in newglob is if the user hasn't modified the default destination folder. If that's the case we should maintain the invariant, stated in help/README, that the default folder should be `./keep`

I think that would make sense, for instance running riv in `/riv/src/program` initially and going to `~/Pictures` and hitting `move` it'd be really weird if we moved the image to `/riv/src/program/keep` rather than `~/Pictures/keep` I personally would think that's unexpected.

I think that we should probably keep track if the user changes the destination folder in case of the instance where the user actually wants the folder name "keep" in a directory. That seems like a bug/unintended behavior.

gurgalex reviewed (changes requested):
@NickHackman 
Thanks for all the help.

Should be the last change request from me.

* Rename paths.`current_dir` to `base_dir`
* Keep selected destination folder the same even if a new glob no longer contains the destination folder.

Example

Selected df
`/home/alex/pictures`

Current glob
`/home/alex/*.png`

New glob
`/home/alex/Downloads/*.png`

Destination folder is still
`/home/alex/pictures`

NickHackman committed:
Paths::curent_dir -> base_dir, better track of df

* Changes to dest_folder is now kept track for the case
of newglob to maintain the invariant that if the user hasn't set dest
folder, images should be moved to "./keep" by default unless specified
otherwise

Davejkane commented:
I like the behaviour of the keep folder as it is right now. As a user I'll probably make a keep folder on my desktop and jump around in different folders using new glob, and it's nice that the keep folder doesn't move with me. 

I've just got a last little bit of QA to do before my blessing.

NickHackman commented:
@Davejkane okay, I'll remove that

Davejkane commented:
In the readme, it says that the command `? or help` will toggle fullscreen. Please correct that. 

~I'm also a little surprised that toggling the help bar with h, and toggling it with the command mode aren't linked, so one can't undo the other. Also, the plain `h` command no longer works when there are no images loaded on initial load, even if I then newglob to get a bunch of pictures, h doesn't work.~

Scrap that, hadn't seen you fixed it to question mark. 👍

Davejkane reviewed (approved):
You did good. Well done. That was a bit epic, but we've pulled through. If there are any bugs, we'll catch them in follow up tickets, but it got past my QA for now. Now that we have this functionality, I'm going to write a couple of tickets.

gurgalex commented on README.md:
```diff
@@ -70,6 +70,19 @@ Set the maximum number of images to be displayed `m` or `--max` flag. 0 means in
+| Command                     | Action                                                   |
+|-----------------------------|----------------------------------------------------------|
+| ng OR newglob [glob]        | **Required argument** the new glob/directory/file        |
+| ? OR help                   | Toggle fullscreen mode                                   |
```
Toggle help menu?

Davejkane replied:
fuck that, I can fix it directly. Let's just get this merged.

Davejkane commented:
@gurgalex are you happy with this PR now? (please approve if so)

gurgalex commented:
@Davejkane 

Doing some testing to see if this is functioning.

>I like the behaviour of the keep folder as it is right now. As a user I'll probably make a keep folder on my desktop and jump around in different folders using new glob, and it's nice that the keep folder doesn't move with me.

gurgalex commented:
Seems fine to me. Even though I'm not seeing a commit that undid the changes for a moving `./keep` directory.

Davejkane commented:
I'll do the merge.
